Subclass existentials - PrintAsObjC support

slavapestov opened the pull request.

slavapestov committed:
PrintAsObjC: Support for subclass existentials

slavapestov commented:
@swift-ci Please smoke test

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1579,62 +1580,65 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+        if (CD->isObjC())
+          os << "SWIFT_METATYPE(" << getNameForObjC(CD) << ")";
+        else
+          os << "Class";
```
When can this happen?

slavapestov replied:
Probably never, o was cargo culting the Metatype case where we seem to print non-@objc metatypes as Class.

jrose-apple replied:
Ah, I didn't remember that was there. I suspect it's not doing anything there either at this point, but it's fine to not investigate that right now.

DougGregor commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1579,62 +1580,65 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+        os << getNameForObjC(CD);
+        if (auto *BGT = layout.superclass->getAs<BoundGenericClassType>()) {
+          os << '<';
+          print(BGT->getGenericArgs()[0], None);
```
You can use the `interleave` algorithm here...

slavapestov replied:
Ah, good idea. This was a bit of copy and paste on my part so I cleaned it up and factored it out into a new method.

DougGregor reviewed (approved):
LGTM

slavapestov committed:
PrintAsObjC: We don't need to print metatypes of non-@objc classes

At least, none of the existing tests hit this case.

slavapestov committed:
PrintAsObjC: Drop redundant protocols from inheritance clauses

There was a FIXME for that. Seems easy enough to address,
please shout if this breaks anything.

slavapestov committed:
PrintAsObjC: Tighten some screws

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test
